Fix base types of second paramters used in Avx.PermuteVar and Avx2.Permute8x32 for float, double overloads

4creators opened the pull request:
Fixes #19579 

cc @CarolEidt @eerhardt @fiigii @tannergooding

CarolEidt commented:
The changes look good, but it seems that the tests will need to be updated as well.

4creators commented:
> The changes look good, but it seems that the tests will need to be updated as well.

Working on this

fiigii commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs:
```diff
@@ -761,22 +761,22 @@ public abstract class Avx : Sse42
         ///   VPERMILPS xmm, xmm, xmm/m128
         /// </summary>
-        public static Vector128<float> PermuteVar(Vector128<float> left, Vector128<float> mask) { throw new PlatformNotSupportedException(); }
+        public static Vector128<float> PermuteVar(Vector128<float> left, Vector128<int> mask) { throw new PlatformNotSupportedException(); }
```
Please change the parameter name to `control` that is not a "mask".

fiigii commented:
>  but it seems that the tests will need to be updated as well.

You probably need to disable (remove) the current tests and update them after CoreFX change refect to CoreCLR repo.

fiigii commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -1294,7 +1294,7 @@ public abstract class Avx2 : Avx
         ///   VPERMPS ymm, ymm/m256, imm8
         /// </summary>
-        public static Vector256<float> PermuteVar8x32(Vector256<float> left, Vector256<float> mask) => PermuteVar8x32(left, mask);
+        public static Vector256<float> PermuteVar8x32(Vector256<float> left, Vector256<int> mask) => PermuteVar8x32(left, mask);
```
Please change the parameter name to control that is not a "mask".

tannergooding commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs:
```diff
@@ -761,22 +761,22 @@ public abstract class Avx : Sse42
         ///   VPERMILPS xmm, xmm, xmm/m128
         /// </summary>
-        public static Vector128<float> PermuteVar(Vector128<float> left, Vector128<float> mask) { throw new PlatformNotSupportedException(); }
+        public static Vector128<float> PermuteVar(Vector128<float> left, Vector128<int> mask) { throw new PlatformNotSupportedException(); }
```
Why? I thought that, elsewhere, we had explicitly kept these the same type (and we have had several PRs moving these off of the integer types and onto the same type as the first parameter)....

This definitely needs to be discussed during the next design-review session. CC. @eerhardt, @terrajobst

fiigii replied:
We only keep the same base type for mask vectors, but here is actually not mask. That is why I suggest to rename to “control”.

fiigii replied:
You can see some intrinsic that takes “index” vector always with integer base type.

tannergooding replied:
Ah, I see, 👍 then.

4creators commented:
Updated PR, addressed code review feedback and removed conflicting tests. Adding tests after related changes in corefx repo are absorbed is tracked by #19624

tannergooding commented:
@4creators, could you get the corresponding CoreFX PR up and ready. Once that is done, we should be good to merge.

4creators commented:
@tannergooding Done PTAL

tannergooding commented:
Thanks for the work!
